File: packages/sokomind-ui/src/solver/verification.ts
import type { Direction, SolutionStep, SolverResult } from "./types.ts";
import { isSolved } from "./types.ts";

export interface BoardCell {
  row: number;
  col: number;
}

export interface ParsedBoard {
  walls: Set<string>;
  goals: Array<{ position: BoardCell; label: string }>;
  initialRobot: BoardCell;
  initialBoxes: Array<{ id: string; label: string; position: BoardCell }>;
}

interface ReplayState {
  robot: BoardCell;
  boxes: Array<{ id: string; label: string; position: BoardCell }>;
  moves: number;
  pushes: number;
}

function cellKey(c: BoardCell): string {
  return `${c.row},${c.col}`;
}

function offset(pos: BoardCell, dir: Direction): BoardCell {
  switch (dir) {
    case "Up":
      return { row: pos.row - 1, col: pos.col };
    case "Down":
      return { row: pos.row + 1, col: pos.col };
    case "Left":
      return { row: pos.row, col: pos.col - 1 };
    case "Right":
      return { row: pos.row, col: pos.col + 1 };
  }
}

function stepReplay(
  board: ParsedBoard,
  state: ReplayState,
  dir: Direction,
): { state: ReplayState; moved: boolean; pushed: boolean } {
  const target = offset(state.robot, dir);

  if (board.walls.has(cellKey(target))) {
    return { state, moved: false, pushed: false };
  }

  const boxIdx = state.boxes.findIndex(
    (b) => b.position.row === target.row && b.position.col === target.col,
  );

  if (boxIdx >= 0) {
    const pushTarget = offset(target, dir);

    if (board.walls.has(cellKey(pushTarget))) {
      return { state, moved: false, pushed: false };
    }
    if (
      state.boxes.some(
        (b) =>
          b.position.row === pushTarget.row &&
          b.position.col === pushTarget.col,
      )
    ) {
      return { state, moved: false, pushed: false };
    }

    const newBoxes = state.boxes.map((b, i) =>
      i === boxIdx ? { ...b, position: pushTarget } : { ...b },
    );

    return {
      state: {
        robot: target,
        boxes: newBoxes,
        moves: state.moves + 1,
        pushes: state.pushes + 1,
      },
      moved: true,
      pushed: true,
    };
  }

  return {
    state: {
      robot: target,
      boxes: state.boxes,
      moves: state.moves + 1,
      pushes: state.pushes,
    },
    moved: true,
    pushed: false,
  };
}

function checkSolved(board: ParsedBoard, boxes: ReplayState["boxes"]): boolean {
  return boxes.every((box) =>
    board.goals.some(
      (g) =>
        g.position.row === box.position.row &&
        g.position.col === box.position.col &&
        g.label === box.label,
    ),
  );
}

export function parseBoardFromRows(rows: string[]): ParsedBoard {
  const walls = new Set<string>();
  const goals: ParsedBoard["goals"] = [];
  const initialBoxes: ParsedBoard["initialBoxes"] = [];
  let initialRobot: BoardCell = { row: 0, col: 0 };
  let boxId = 0;

  for (let r = 0; r < rows.length; r++) {
    const row = rows[r]!;
    for (let c = 0; c < row.length; c++) {
      const ch = row[c]!;
      if (ch === "O") {
        walls.add(cellKey({ row: r, col: c }));
      } else if (ch === "R") {
        initialRobot = { row: r, col: c };
      } else if (ch === "X") {
        initialBoxes.push({
          id: `box-${boxId++}`,
          label: "X",
          position: { row: r, col: c },
        });
      } else if (ch === "S") {
        goals.push({
          position: { row: r, col: c },
          label: "X",
        });
      } else if (ch >= "A" && ch <= "Z" && ch !== "O" && ch !== "R" && ch !== "S" && ch !== "X") {
        initialBoxes.push({
          id: `box-${boxId++}`,
          label: ch,
          position: { row: r, col: c },
        });
      } else if (ch >= "a" && ch <= "z") {
        goals.push({
          position: { row: r, col: c },
          label: ch.toUpperCase(),
        });
      }
    }
  }

  return { walls, goals, initialRobot, initialBoxes };
}

export interface VerificationResult {
  valid: boolean;
  moves: number;
  pushes: number;
  solved: boolean;
  error?: string;
}

export function verifySolution(
  rows: string[],
  steps: SolutionStep[],
): VerificationResult {
  const board = parseBoardFromRows(rows);

  let state: ReplayState = {
    robot: board.initialRobot,
    boxes: board.initialBoxes.map((b) => ({ ...b })),
    moves: 0,
    pushes: 0,
  };

  for (let i = 0; i < steps.length; i++) {
    const step = steps[i]!;
    const { state: next, moved, pushed } = stepReplay(
      board,
      state,
      step.direction,
    );

    if (!moved) {
      return {
        valid: false,
        moves: state.moves,
        pushes: state.pushes,
        solved: false,
        error: `step ${i}: move ${step.direction} is blocked`,
      };
    }

    if (step.pushed && !pushed) {
      return {
        valid: false,
        moves: state.moves,
        pushes: state.pushes,
        solved: false,
        error: `step ${i}: expected push but no box was pushed`,
      };
    }

    state = next;
  }

  const solved = checkSolved(board, state.boxes);

  return {
    valid: true,
    moves: state.moves,
    pushes: state.pushes,
    solved,
    error: solved ? undefined : "replay completed but puzzle not solved",
  };
}

export function verifyResult(
  rows: string[],
  result: SolverResult,
): VerificationResult {
  if (!isSolved(result.status) || !result.solution) {
    return {
      valid: false,
      moves: 0,
      pushes: 0,
      solved: false,
      error: "no solution to verify",
    };
  }

  return verifySolution(rows, result.solution.steps);
}

File: crates/sokomind-solver/src/verification.rs
use sokomind_core::board::ParsedBoard;
use sokomind_core::game::{create_snapshot, step_snapshot};
use sokomind_core::position::Direction;

/// Verify a solution by replaying it through the core game engine.
/// Returns Ok(moves, pushes) or Err with the step index that failed.
pub fn verify_solution(
    board: &ParsedBoard,
    steps: &[(Direction, bool)],
) -> Result<(u32, u32), VerificationError> {
    let mut snapshot = create_snapshot(board);

    for (i, &(dir, expected_push)) in steps.iter().enumerate() {
        let transition = step_snapshot(board, &snapshot, dir);
        if !transition.moved {
            return Err(VerificationError::Blocked { step: i });
        }
        if transition.pushed != expected_push {
            return Err(VerificationError::PushMismatch {
                step: i,
                expected: expected_push,
                actual: transition.pushed,
            });
        }
        snapshot = transition.snapshot;
    }

    if !snapshot.solved {
        return Err(VerificationError::NotSolved);
    }

    Ok((snapshot.moves, snapshot.pushes))
}

#[derive(Debug, PartialEq, Eq)]
pub enum VerificationError {
    Blocked {
        step: usize,
    },
    PushMismatch {
        step: usize,
        expected: bool,
        actual: bool,
    },
    NotSolved,
}

impl std::fmt::Display for VerificationError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            VerificationError::Blocked { step } => write!(f, "blocked at step {}", step),
            VerificationError::PushMismatch {
                step,
                expected,
                actual,
            } => write!(
                f,
                "push mismatch at step {}: expected={}, actual={}",
                step, expected, actual
            ),
            VerificationError::NotSolved => write!(f, "puzzle not solved after replay"),
        }
    }
}

impl std::error::Error for VerificationError {}

#[cfg(test)]
mod tests {
    use super::*;
    use sokomind_core::board::parse_board;

    #[test]
    fn verify_valid_solution() {
        // OOOOO
        // O  SO
        // O XRO  (box at (2,2), goal at (1,3), robot at (2,3))
        // O   O
        // OOOOO
        // Solution: robot goes up (to 1,3), left (pushes box left to 2,1? no)
        // Actually: push box up (robot at 2,3 goes left to 2,2 pushing box to 2,1... no)
        // Let me think: robot at (2,3), box at (2,2), goal at (1,3)
        // Push box up: robot needs to be below box at (3,2), push up moves box to (1,2)
        // Then push right: robot needs to be left of box at (1,1), push right moves box to (1,3)
        // Robot path: (2,3) → D(3,3) → L(3,2) → L(3,1) → U(2,1) → U(1,1)
        // wait, we need to get robot below box first
        // Robot at (2,3). Box at (2,2).
        // D to (3,3), L to (3,2): robot below box. U pushes box from (2,2) to (1,2).
        // Now box at (1,2), robot at (2,2).
        // Need robot left of box: D to (3,2), L to (3,1), U to (2,1), U to (1,1).
        // R pushes box from (1,2) to (1,3). Solved!
        let rows = &["OOOOO", "O  SO", "O XRO", "O   O", "OOOOO"];
        let board = parse_board(rows).unwrap();

        let steps = vec![
            (Direction::Down, false), // (3,3)
            (Direction::Left, false), // (3,2)
            (Direction::Up, true),    // push box (2,2)->(1,2), robot to (2,2)
            (Direction::Down, false), // (3,2)
            (Direction::Left, false), // (3,1)
            (Direction::Up, false),   // (2,1)
            (Direction::Up, false),   // (1,1)
            (Direction::Right, true), // push box (1,2)->(1,3), robot to (1,2). Solved!
        ];

        let result = verify_solution(&board, &steps);
        assert!(result.is_ok(), "expected valid: {:?}", result);
        let (moves, pushes) = result.unwrap();
        assert_eq!(moves, 8);
        assert_eq!(pushes, 2);
    }

    #[test]
    fn verify_blocked() {
        let rows = &["OOOOO", "O  SO", "O XRO", "O   O", "OOOOO"];
        let board = parse_board(rows).unwrap();
        let steps = vec![(Direction::Right, false)]; // into wall
        assert_eq!(
            verify_solution(&board, &steps),
            Err(VerificationError::Blocked { step: 0 })
        );
    }

    #[test]
    fn verify_not_solved() {
        let rows = &["OOOOO", "O  SO", "O XRO", "O   O", "OOOOO"];
        let board = parse_board(rows).unwrap();
        let steps = vec![(Direction::Down, false)]; // just walk, not solved
        assert_eq!(
            verify_solution(&board, &steps),
            Err(VerificationError::NotSolved)
        );
    }
}
